internal/engine/marein/marein.go

```diff
@@ -0,0 +1,167 @@
+package engine_marein
+
+import (
+	"math"
+	"math/rand"
+
+	"github.com/gaming-platform/connect-four-bot/internal/connectfour"
+)
+
+// A custom implementation with some heuristics.
+// No big algorithms, favors readability over efficiency, just for fun.
+// It currently ignores depth > 1.
+
+func CreateCalculateNextMove(maxDepth int) func(game *connectfour.Game) (int, bool) {
+	return func(game *connectfour.Game) (int, bool) {
+		return calculateNextMove(game, maxDepth)
+	}
+}
+
+func calculateNextMove(game *connectfour.Game, maxDepth int) (int, bool) {
+	current, opponent := game.GetCurrentPlayerColors()
+
+	freeColumns := game.GetFreeColumns()
+	if len(freeColumns) == 0 {
+		return 0, false
+	}
+
+	// Win if possible.
+	if x, ok := findWinningMove(game, freeColumns, current); ok {
+		return x, true
+	}
+
+	// Prevent opponent from winning.
+	if x, ok := findWinningMove(game, freeColumns, opponent); ok {
+		return x, true
+	}
+
+	// Prevent opponent from creating a fork.
+	if x, ok := findForkingMove(game, freeColumns, opponent); ok {
+		return x, true
+	}
+
+	// Create a fork if possible.
+	if x, ok := findForkingMove(game, freeColumns, current); ok {
+		return x, true
+	}
+
+	// Prevent forcing winning sequence in the future.
+	if x, ok := findMoveForcingWinningSequence(game, freeColumns, opponent); ok {
+		return x, true
+	}
+
+	// Set up a forcing winning sequence in the future.
+	if x, ok := findMoveForcingWinningSequence(game, freeColumns, current); ok {
+		return x, true
+	}
+
+	// Find a random move that does not allow opponent to win next turn.
+	filteredColumns := freeColumns
+	for {
+		x := findRandomLegalMoveThatPrefersCenter(game, filteredColumns)
+		y, _ := game.NextFreeRow(x)
+		gameClone := game.Clone()
+		gameClone.ApplyMove(x, y)
+
+		if !findThreat(gameClone, maxDepth) {
+			return x, true
+		}
+
+		filteredColumns = removeColumn(filteredColumns, x)
+		if len(filteredColumns) == 0 { // Threats everywhere, just move somewhere.
+			return freeColumns[0], true
+		}
+	}
+}
+
+func findWinningMove(game *connectfour.Game, freeColumns []int, color int) (int, bool) {
+	for _, x := range freeColumns {
+		y, _ := game.NextFreeRow(x)
+		gameClone := game.Clone()
+		gameClone.ForceMove(x, y, color)
+
+		if connectfour.IsWinningMove(gameClone, x, y, color) {
+			return x, true
+		}
+	}
+
+	return 0, false
+}
+
+// findForkingMove looks for positions that will result in: 2 0 1 1 1 0 2.
+func findForkingMove(game *connectfour.Game, freeColumns []int, color int) (int, bool) {
+	for _, x := range freeColumns {
+		y, _ := game.NextFreeRow(x)
+		gameClone := game.Clone()
+		gameClone.ForceMove(x, y, color)
+
+		nextFreeColumns := gameClone.GetFreeColumns()
+		if firstWinX, ok := findWinningMove(gameClone, nextFreeColumns, color); ok {
```

Comment:
Should check for threats too, e.g. `[{"x":4,"y":6,"color":1},{"x":4,"y":5,"color":2},{"x":4,"y":4,"color":1},{"x":3,"y":6,"color":2},{"x":4,"y":3,"color":1},{"x":5,"y":6,"color":2},{"x":4,"y":2,"color":1},{"x":4,"y":1,"color":2},{"x":5,"y":5,"color":1},{"x":5,"y":4,"color":2},{"x":5,"y":3,"color":1},{"x":6,"y":6,"color":2},{"x":3,"y":5,"color":1},{"x":2,"y":6,"color":2},{"x":5,"y":2,"color":1},{"x":3,"y":4,"color":2},{"x":3,"y":3,"color":1},{"x":3,"y":2,"color":2},{"x":3,"y":1,"color":1},{"x":2,"y":5,"color":2},{"x":2,"y":4,"color":1},{"x":2,"y":3,"color":2}]`.
Shouldn't have moved into column 2. This example should be a test case.